### PR TITLE
Remove restriction on subgroups in version catalogs for plugins

### LINKED ISSUE #18142: Version Catalog: Sub-accessors and alias overload not supported for plugins.
### Expected Behavior
I have plugin definitions `kotlin-js` and `kotlin-js-action`.  As of 7.2, this works fine for libraries, and I expected it to work for plugins, too.

### Current Behavior
I get:
```
Script compilation error:

  Line 5:     alias(libs.plugins.kotlin.js) apply false
                    ^ Type mismatch: inferred type is LibrariesForLibs.KotlinJsPluginAccessors! but Provider<PluginDependency!>! was expected
```

### Context
This is an annoyance whenever you want to use plugins where one has an ID containing the other's ID.  This likely won't be very common, but I would think it's doable since it's done for dependencies.

### Steps to Reproduce 
Use `libs.versions.toml`:
```toml
[versions]
kotlin = "1.5.21"
kotlin-js-action = "1.4.3"

[plugins]
kotlin-js = { id = "org.jetbrains.kotlin.js", version.ref = "kotlin" }
kotlin-js-action = { id = "com.github.rnett.ktjs-github-action", version.ref = "kotlin-js-action" }
```

Use `build.gradle.kts`:
```kotlin
plugins {
    alias(libs.plugins.kotlin.jvm) apply false
    alias(libs.plugins.kotlin.js) apply false
}
```

### Your Environment
Windows, Gradle 7.2.

### PR DESCRIPTION
**Pull request by asodja**
This removes a restriction on subgroups in version catalogs for plugins. Fix for libraries was already applied before but for plugins this did not work.

This means, that:
```
plugins {
    alias(libs.plugins.greeter)
    alias(libs.plugins.greeter.second)
}
```
now works.

Note that I had to move `ProviderConvertible<T>` to public API.

Fixes #18142

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test QFL

File: subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package org.gradle.api.provider;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.NonExtensible;
+import org.gradle.internal.HasInternalProtocol;
+
+/**
+ * An object that can be transformed to {@link Provider<T>}.
```
Comment (ljacomet) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java:24]: 💄 : `An object that can be converted to a {@link Provider<T>}`

File: subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * @param <T> Type of value represented by provider
+ * @since 7.3
+ */
+@Incubating
+@HasInternalProtocol
+@NonExtensible
+public interface ProviderConvertible<T> {
+
+    /**
+     * Returns a new {@link Provider<T>} from this object.
```
Comment (ljacomet) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java:35]: 💄 : I would not use "new" in this context. Calling this method twice should actually return the same instance.

**Review by ljacomet — COMMENTED** [review]
LGTM overall 👍 

Minor nitpicks on the public API Javadoc though:

File: subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java (hunk 1 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java]: Fixed.

File: subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java (hunk 2 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderConvertible.java]: Fixed.

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test this

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/CatalogPluginsGroovyDSLIntegrationTest.groovy [hunk 3]
```diff
@@ -187,4 +187,84 @@ dependencyResolutionManagement {
...
+
+        when:
+        succeeds(firstLevelTask, secondLevelPluginTask)
+
+        then:
+        outputContains 'Hello from first plugin!'
+        outputContains 'Hello from second plugin!'
+    }
+
+    def "can apply a plugin via buildscript that has sub-accessors"() {
```
Comment (ljacomet) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/CatalogPluginsGroovyDSLIntegrationTest.groovy:221]: Looking some more into the added tests, I feel like this one is a superset of "can apply a plugin declared in a catalog via buildscript".
So maybe that other one can be removed. Wdyt?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/CatalogPluginsKotlinDSLIntegrationTest.groovy [hunk 4]
```diff
@@ -234,4 +234,92 @@ dependencyResolutionManagement {
...
+
+        when:
+        succeeds(firstLevelTask, secondLevelPluginTask)
+
+        then:
+        outputContains 'Hello from first plugin!'
+        outputContains 'Hello from second plugin!'
+    }
+
+    def "can apply a plugin via buildscript that has sub-accessors"() {
```
Comment (ljacomet) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/CatalogPluginsKotlinDSLIntegrationTest.groovy:277]: Same comment as for the groovy one.

**Review by ljacomet — APPROVED** [review]
LGTM!

Minor comments on possible tests duplication though - see below

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/CatalogPluginsGroovyDSLIntegrationTest.groovy (hunk 3 above)
Comment (asodja) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog/CatalogPluginsGroovyDSLIntegrationTest.groovy:221]: Good point. I will delete the other one

**Comment by asodja** [comment]
@bot-gradle test and merge

**Comment by asodja** [comment]
@bot-gradle test and merge

**Pull request merged**